cpp/src/ffi_go/ir/serializer.h

```diff
@@ -0,0 +1,109 @@
+#ifndef FFI_GO_IR_SERIALIZER_H
+#define FFI_GO_IR_SERIALIZER_H
+// header must support C, making modernize checks inapplicable
+// NOLINTBEGIN(modernize-deprecated-headers)
+// NOLINTBEGIN(modernize-use-trailing-return-type)
+
+#include <stdint.h>
+#include <stdlib.h>
+
+#include <ffi_go/defs.h>
+
+/**
+ * Clean up the underlying ir::Serializer of a Go ir.Serializer.
+ * @param[in] ir_serializer Address of a ir::Serializer created and returned by
+ *     ir_serializer_serialize_*_preamble
+ */
+void ir_serializer_close(void* ir_serializer);
+
+/**
+ * Given the fields of a CLP IR premable, serialize them into an IR byte stream
+ * with eight byte encoding. An ir::Serializer will be allocated to use as the
+ * backing storage for a Go ir.Serializer (i.e. subsequent calls to
+ * ir_serializer_serialize_*_log_event). All pointer parameters must be non-null
+ * (non-nil Cgo C.<type> pointer or unsafe.Pointer from Go).
+ * @param[in] ts_pattern Format string for the timestamp to be used when
+ *     deserializing the IR
+ * @param[in] ts_pattern_syntax Type of the format string for understanding how
+ *     to parse it
+ * @param[in] time_zone_id TZID timezone of the timestamps in the IR
+ * @param[out] ir_serializer_ptr Address of a new ir::Serializer
+ * @param[out] ir_view View of a IR buffer containing the serialized preamble
+ * @return ffi::ir_stream::IRErrorCode forwared from
+ *     ffi::ir_stream::eight_byte_encoding::encode_preamble
+ */
+int ir_serializer_serialize_eight_byte_preamble(
+        StringView ts_pattern,
+        StringView ts_pattern_syntax,
+        StringView time_zone_id,
+        void** ir_serializer_ptr,
+        ByteSpan* ir_view
+);
+
+/**
+ * Given the fields of a CLP IR premable, serialize them into an IR byte stream
+ * with four byte encoding. An ir::Serializer will be allocated to use as the
+ * backing storage for a Go ir.Serializer (i.e. subsequent calls to
+ * ir_serializer_serialize_*_log_event). All pointer parameters must be non-null
+ * (non-nil Cgo C.<type> pointer or unsafe.Pointer from Go).
+ * @param[in] ts_pattern Format string for the timestamp to be used when
+ *     deserializing the IR
+ * @param[in] ts_pattern_syntax Type of the format string for understanding how
+ *     to parse it
+ * @param[in] time_zone_id TZID timezone of the timestamps in the IR
+ * @param[out] ir_serializer_ptr Address of a new ir::Serializer
+ * @param[out] ir_view View of a IR buffer containing the serialized preamble
+ * @return ffi::ir_stream::IRErrorCode forwared from
+ *     ffi::ir_stream::four_byte_encoding::encode_preamble
+ */
+int ir_serializer_serialize_four_byte_preamble(
```

Comment:
I think the naming of these two functions (`ir_serializer_serialize_four_byte_preamble` and `ir_serializer_serialize_eight_byte_preamble`) is unclear. Essentially, these functions not just serialize the preamble bytes, but creating a new serializer by allocating heap memory + initialization. We should be more explicit to call it with sth like `ir_serializer_create_xxxx_byte_serializer_with_preamble`